### ISSUE TITLE
Share RenderBackend between multiple Renderer/RenderAPIs

### ISSUE THREAD

**Issue by kvark** [servo/webrender#1350]
It would be great to investigate if we gain anything from sharing a backend between different GL contexts (that are associated with different `Renderer` and `RenderApi` instances. At the very minimal, we could share the rayon thread pool, but there can be more.

cc @nical @jrmuizel @sotaroikeda

**Comment by nical** [comment]
Sharing the rayon thread pool is already possible without further modification of webrender, see the [gecko patch to do that](https://bugzilla.mozilla.org/show_bug.cgi?id=1370879) for example.

I would love to at least share the render backend thread between webrender instances.
Going further and sharing resources like fonts and rasterized glyphs would be really nice too.

I doubt we can have a single webrender instance for all windows, due to the way windows and their gl contexts are tied together but i'd like to be proven wrong.

**Comment by jrmuizel** [comment]
See also https://github.com/servo/webrender/issues/1233

**Comment by nical** [comment]
> See also #1233

My understanding is that this was because servo is using a wr instance per window (just like gecko currently does) and gl make_current wasn't properly used, causing a window to send commands to the other window's gl context. We already have the same form of multi-window support.

**Comment by kvark** [comment]
> Sharing the rayon thread pool is already possible without further modification of webrender, see the gecko patch to do that for example.

Right, I'm aware of the patch. The issue can be re-formulated as "what parts of WR we can share between different contexts", and I brought up the thread pool as a sort of lower bound for that metric, that we already achieved.

From the first look at the backend:
  - `GpuCache`, `TextureCache`, `webgl_textures` - would only work if we share textures between GL contexts
  - `font_templates`, `image_templates`, `cached_glyphs`, `cached_images`, `cached_glyph_dimensions` - should be already possible to share
  - `blob_image_renderer` - I suppose can be shared as well
  - `result_tx` - we could just provide the sender in the API request instead of storing it
  - `Scene` - appears to be sharable? given that it contains all the pipelines for all contexts, and we temporarily switch the root pipeline to the current document
  - `Frame` - doesn't appear to be sharable

**Comment by nical** [comment]
Out of order:

 - Blob image renderer doesn't hold much state (it uses the shared thread pool) and only produces data that is useful to the specific scene it renders, so there isn't much gain from sharing it.

 - Sharing the render backend thread (and have several render backend objects on it) should not be too hard and we should really make it possible. It would also make the other items below a lot easier.

 - Font templates could be shared, raw fonts are stored as arcs so we could have a global font template store, and each window's font cache would just refer to fonts from this global store. I think that this should not be too hard (if we already share the render backend thread) to implement and would avoid storing the font data twice (at least for windows that come from the same content process).

 - Sharing rasterized glyphs would be really cool. Harder to do, but I would be surprised if we couldn't get some reuse out of this, if only for the windows of the browser UI which certainly use mostly the same fonts and sizes.

 - The glyph rasterizer itself is not shared at the moment. It uses the shared thread pool, but each glyph rasterizer has its own set of font contexts (1 per worker thread, + 1). We could make it so we share the font contexts as well, this would only be useful if we are already able to share the font templates.

 - For image templates, we could share some of them similarly to font templates (stored as arcs). I don't know how hard it would be to recognize that an image is the same in two windows (or even tabs), because as we delve into the gfx side of things we tend to lose information about where the data came from.

 - The scenes should be different for each window, so sharing them would not provide a lot of benefits as far as I understand.


Sharing resources between gl contexts has proven to be very buggy so I'd rather not unless there are huge benefits. We'll already have plenty of driver bugs on our hands. I say gl context but the same applies to DXGI unfortunately (less so, but way too much still).
